contracts/modules/metadata/CoreMetadataModule.sol

```diff
@@ -0,0 +1,114 @@
+            revert Errors.CoreMetadataModule__MetadataAlreadyFrozen();
+        }
+        _;
+    }
+
+    /// @notice Creates a new CoreMetadataModule instance.
+    /// @param accessController The address of the AccessController contract.
+    /// @param ipAccountRegistry The address of the IPAccountRegistry contract.
+    constructor(
+        address accessController,
+        address ipAccountRegistry
+    ) AccessControlled(accessController, ipAccountRegistry) {}
+
+    /// @notice Update the nftTokenURI for an IP asset,
+    /// by retrieve the latest TokenURI from IP NFT to which the IP Asset bound.
+    /// @dev Can only be called once per IP asset to prevent overwriting.
+    /// @param ipId The address of the IP asset.
+    /// @param nftMetadataHash A bytes32 hash representing the metadata of the NFT.
+    /// This metadata is associated with the IP Asset and is accessible via the NFT's TokenURI.
+    /// Use bytes32(0) to indicate that the metadata is not available.
+    function updateNftTokenURI(address ipId, bytes32 nftMetadataHash) external verifyPermission(ipId) {
+        _updateNftTokenURI(ipId, nftMetadataHash);
+    }
+
+    /// @notice Sets the metadataURI for an IP asset.
+    /// @dev Can only be called once per IP asset to prevent overwriting.
+    /// @param ipId The address of the IP asset.
+    /// @param metadataURI The metadataURI to set for the IP asset.
+    /// @param metadataHash The hash of metadata at metadataURI.
+    /// Use bytes32(0) to indicate that the metadata is not available.
+    function setMetadataURI(
+        address ipId,
+        string memory metadataURI,
+        bytes32 metadataHash
+    ) external verifyPermission(ipId) {
+        _setMetadataURI(ipId, metadataURI, metadataHash);
+    }
+
+    /// @notice Sets all core metadata for an IP asset.
+    /// @dev Can only be called once per IP asset to prevent overwriting.
+    /// @param ipId The address of the IP asset.
+    /// @param metadataURI The metadataURI to set for the IP asset.
+    /// @param metadataHash The hash of metadata at metadataURI.
+    /// Use bytes32(0) to indicate that the metadata is not available.
+    /// @param nftMetadataHash A bytes32 hash representing the metadata of the NFT.
+    /// This metadata is associated with the IP Asset and is accessible via the NFT's TokenURI.
+    /// Use bytes32(0) to indicate that the metadata is not available.
+    function setAll(
+        address ipId,
+        string memory metadataURI,
+        bytes32 metadataHash,
+        bytes32 nftMetadataHash
+    ) external verifyPermission(ipId) {
+        _updateNftTokenURI(ipId, nftMetadataHash);
+        _setMetadataURI(ipId, metadataURI, metadataHash);
+    }
+
+    /// @notice make all metadata of the IP Asset immutable.
+    /// @param ipId The address of the IP asset.
+    function freezeMetadata(address ipId) external verifyPermission(ipId) {
+        IIPAccount(payable(ipId)).setBool("IMMUTABLE", true);
```

Comment:
"IMMUTABLE" for metadata immutability might be too broad, could we use "IMMUTABLE_METADATA"?

However, if there are no other use cases for the tag "IMMUTABLE," then using it for metadata looks good.

Reply:
Right. in metadata module, the "IMMUTABLE" flag are only used for metadata.